### ISSUE TITLE
Error when using threads: Target thread must not be current thread

### ISSUE THREAD

**Issue by johnhampton** [kapost/circuitry#51]
A subscriber that uses threads as the async strategy raises the error `Error processing message fefbd69d-91e4-451e-943a-cad8025d4016: Target thread must not be current thread`.

My best guess is that the problem is with the Flush middleware. The way the code is organized the middleware is run asynchronously in the same thread as each individual message. The call to Threader.flush then attempts to join all threads (including itself). Even if one were to make sure to avoid joining the current thread (something like `pool.reject {|t| t == Thread.current}.each(&:join)`, I'm not sure that the code is working as intended. It looks like the worker threads would wait on each other to finish, but there is nothing from stopping the main thread from continuing and polling for more messages before the worker threads complete. 

Assuming the intent is to pull a batch of messages and have all those messages processed before retrieving the next batch, the flush would need be called in the main thread. Something along the lines of: 

```ruby
    def poll(&block)
      poller = Aws::SQS::QueuePoller.new(queue, client: sqs)

      poller.before_request do |_stats|
        throw :stop_polling unless subscribed?
      end

      poller.poll(max_number_of_messages: batch_size, wait_time_seconds: wait_time, skip_delete: true) do |messages|
        process_messages(Array(messages), &block)
        
        # Wait unit all messages have been processed
        Circuitry.flush 
      end
    end
```

I just started using circuitry today so I apologize if I'm way off the mark. The error can be reproduced with the following code:

```ruby
require 'circuitry'
require 'pp'

Circuitry.subscriber_config do |c|
  c.queue_name = ENV['CIRCUITRY_QUEUE_NAME']
  c.access_key = ENV['CIRCUITRY_ACCESS_KEY']
  c.secret_key = ENV['CIRCUITRY_SECRET_KEY']
  c.region = ENV['CIRCUITRY_REGION']
  c.async_strategy = :thread
end

options = {
  lock: true,
  async: :thread,
  timeout: 20,
  wait_time: 10,
  batch_size: 10
}

Circuitry.subscribe(options) do |message, topic_name|
  pp topic_name
  pp message
end
```

results in

```
E, [2016-03-29T22:33:13.490817 #66002] ERROR -- : Error processing message bc8d9d54-109c-4abd-9a7b-fcf8db67be49: Target thread must not be current thread
```

**Comment by brandonc** [comment]
Thanks for your detailed report of the issue. You are correct: it was unintended for flush to join itself in the async strategy. @mhuggins I think we should revert #47 and either flush after batch_size messages are processed as @johnhampton suggests or rely on implementers to flush.

**Comment by brandonc** [comment]
@mhuggins One interesting thing about flush is that it doesn't necessarily guarantee that thread subscribers who publish using a batch will have their messages published because Batcher.flush could be called before Threader.flush. Is that correct?

**Comment by mhuggins** [comment]
Thanks for the report!  This looks like a legitimate issue, we'll definitely want to get this fixed.

As a workaround, you can disable the flush middleware:

```ruby
Circuitry.subscriber_config.middleware.remove Middleware::Entries::Flush
```

I'll be in the office soon and will take a closer look then.

**Comment by mhuggins** [comment]
@johnhampton I just created a new branch named [bug/threaded-flush](https://github.com/kapost/circuitry/tree/bug/threaded-flush).  Can you try pointing your project at that and see if it helps?  (Make sure you remove the workaround from my previous comment, if you were using it.)

**Comment by johnhampton** [comment]
@mhuggins Just wanted to get back to you, my project was extremely time sensitive so I had to move forward using a another library. If you want I can still try to take a look at the fix this weekend.

**Comment by mhuggins** [comment]
@johnhampton No worries, and my apologies for the delay in getting back to you.  I'll plan on testing this out and getting a working fix in place.  Thanks!

**Comment by mhuggins** [comment]
Tested the fix for this, and it's working.  I'm going to merge and get a new release out soon.  Thanks again for the report @johnhampton!

**Comment by mhuggins** [comment]
This fix is released as part of 3.1.4.
